### ISSUE TITLE
Spaces at the beginning of lines increase

### ISSUE THREAD

**Issue by TE-0000105050** [Frederick888/external-editor-revived#61]
# Description

It is very happy to use this extension.
But External Editor Revived passes a text file with an extra space character at the beginning of each line that starts with one or more spaces.

For example, a text like below
```
123
 123
```
becomes
```
123
  123
```

This symptom used to occur sometimes before, but it occurs always since I updated Thunderbird from ver. 9x to 10x.

Could you investigate and fix the issue, please?

Thank you!

# Environment

- OS: Windows
- Thunderbird version: 102.3.0 (64 bit)
- ExtEditorR extension version: 0.4.0
- `external-editor-revived --version` output: External Editor Revived native messaging host for windows (x86_64) v0.4.0

# Configuration

Editor: Emacs
Shell: C:\cygwin64\bin\bash.exe

# Logs

```
12:14:24.497 This page is in Quirks Mode. Page layout may be impacted. For Standards Mode use “<!DOCTYPE html>”.MimeMessageUtils.jsm:148:23
12:14:24.501 ExtEditorR sending:  
Object { configuration: {…}, tab: {…}, composeDetails: {…} }
​
composeDetails: Object { from: "foobar@example.com", overrideDefaultFcc: false, type: "draft", … }
​​
additionalFccFolder: ""
​​
attachVCard: false
​​
attachments: Array []
​​
bcc: Array []
​​
body: "<html><head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\"><style></style></head><body style=\"font-family: -moz-fixed; white-space: pre-wrap;\">123<br> 123<br><br></body></html>"
​​
cc: Array []
​​
customHeaders: Array []
​​
deliveryFormat: null
​​
deliveryStatusNotification: false
​​
followupTo: Array []
​​
from: "foobar@example.com"
​​
identityId: "id1"
​​
isPlainText: true
​​
newsgroups: Array []
​​
overrideDefaultFcc: false
​​
overrideDefaultFccFolder: null
​​
plainTextBody: "123\n  123"
​​
priority: "normal"
​​
relatedMessageId: 1
​​
replyTo: Array []
​​
returnReceipt: false
​​
subject: ""
​​
to: Array []
​​
type: "draft"
​​
<prototype>: Object { … }
​
configuration: Object { version: "0.4.0", shell: "C:\\cygwin64\\bin\\bash.exe", template: "C:/Program\\ Files/Emacs/emacs-28.1/bin/emacsclientw.exe /path/to/temp.eml", … }
​
tab: Object { id: 6, index: 0, windowId: 53, … }
​
<prototype>: Object { … }
background.js:115:11
```

# Screenshots

![extra-spaces](https://user-images.githubusercontent.com/40198360/192187408-c8f99fa0-1bf7-4d18-bd49-82147cf42598.png)

**Comment by Frederick888** [comment]
Interesting... I can also reproduce this under Linux using Thunderbird 102.3.0.

FWIW I just grab the original returning value [1] from `compose.getComposeDetails()` API and send it to the messaging host. There is no post-processing and this hasn't changed since the very first version of this project. So if it's been on and off I'm afraid this can be some sort of regression on Thunderbird's side.

I've been a little under the pump recently so I don't have too much time to dig deeper into this right now. But if you can spare some time, I suggest you either give mozregression [2] a shot or at least install a few older Thunderbirds to figure out the starting point of this issue.

[1] https://github.com/Frederick888/external-editor-revived/blob/a13ea62b9c705b319ef8ab1402d5f8f875f16860/extension/background.js#L100
[2] https://mozilla.github.io/mozregression/

**Comment by Frederick888** [comment]
By the way since they apparently changed something here between 91 and 102 under Windows according to your #62, I suggest you test out latest Thunderbird 91 ESR first to see if this happens.

**Comment by Frederick888** [comment]
Ouch, is it just me or trailing spaces at the end of each line are gone as well? :facepalm:

**Comment by Frederick888** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1792551.

Please vote for the ticket if you can.

**Comment by TE-0000105050** [comment]
Hi Frederick,
Thank you for your support under your busy situation.

> Ouch, is it just me or trailing spaces at the end of each line are gone as well? 

I see that issue, too, with Thunderbird 102.3.0.

Furthermore, I've noticed another similar difficulty that the new line control character at the end of the last line of the mail draft is deleted when invoking an editor.  I wonder if it comes from the same root cause as the other two difficulties.

The three difficulties are:
1. Spaces at the beginning of lines increase.
2. Trailing space at the end of lines are deleted.
3. The new line control character at the end of the last line is deleted.

I've compared TB 91.13.1 (the last minor version of 91.x) and 102.3.0.  Please see the screenshots tb-91.13.1.png and tb-102.3.0.png, which are attached.  The all three difficulties don't occur on TB 91.13.1, while the all occur on TB 102.3.0.

> Please vote for the ticket if you can.

I voted it.

> But if you can spare some time, I suggest you either give mozregression [2] a shot or at least install a few older Thunderbirds to figure out the starting point of this issue.

You mentioned a Thunderbird's specific commit in bugzilla.  That means you already identified when the bug is introduced, right?

It seems that the commit changed the way of Thunderbird's getComposeDetails() API to get a plain text from reading the plain text that Thunderbird has to converting an HTML to plain text.  I wonder if the commit didn't tackled the root cause of the original bug [Bug 1692439](https://bugzilla.mozilla.org/show_bug.cgi?id=1692439) but just made a workaround, and messed things, so it should be replaced by some right fix...

Thank you,

**Comment by TE-0000105050** [comment]
![tb-91 13 1](https://user-images.githubusercontent.com/40198360/192736970-55513d3c-efe1-4487-919f-7e4bfad8d413.png)
![tb-102 3 0](https://user-images.githubusercontent.com/40198360/192736990-fcb705b6-f7b3-4a6b-8ef7-96cd3161c72d.png)

**Comment by Frederick888** [comment]
> The three difficulties are:
> 
>   1. Spaces at the beginning of lines increase.
>   2. Trailing space at the end of lines are deleted.
>   3. The new line control character at the end of the last line is deleted.

Third issue was introduced by https://hg.mozilla.org/comm-central/rev/097ff52c68edb0a8a62505a1380a67e0ca6fdbef, and https://hg.mozilla.org/comm-central/rev/b96d9df1eacc2092114ac04dcbed3a6cc4612043 seems to be built on top of this.

> I wonder if the commit didn't tackled the root cause of the original bug [Bug 1692439](https://bugzilla.mozilla.org/show_bug.cgi?id=1692439) but just made a workaround, and messed things, so it should be replaced by some right fix...

I didn't spend enough time on this issue to speak about that... Please feel free to comment on my bugzilla ticket if you've any new findings.

**Comment by TE-0000105050** [comment]
Hi Frederick,

The fix of Thunderbird by the Bugzilla ticket was released publicly as Thunderbird 102.5.1, and all the three issues above seems to have been solved.
Thank you very much!

Best regards,

**Comment by Frederick888** [comment]
Also just got 102.5.1 and can confirm it was fixed :tada:
